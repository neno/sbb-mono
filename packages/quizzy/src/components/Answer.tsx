import React, { memo } from 'react';
import { AnswerProps } from '../models';

const Answer: React.FC<AnswerProps> = memo(
    ({
        id,
        questionId,
        questionType,
        title,
        correct,
        checked,
        toggleAnswer,
        showResults,
    }) => {
        const RadioButton = () => (
            <input
                type="radio"
                name={`answer-${questionId}`}
                id={`answer-${id}`}
                checked={checked}
                onChange={() => toggleAnswer(id)}
                disabled={showResults}
            />
        );
        const Checkbox = () => (
            <input
                type="checkbox"
                name={`answer-${id}`}
                id={`answer-${id}`}
                checked={checked}
                onChange={() => toggleAnswer(id)}
                disabled={showResults}
            />
        );

        let labelClassNames = `a-quiz-answer a-quiz-answer--${questionType}`;
        if (showResults) {
            labelClassNames += correct
                ? ' a-quiz-answer--correct'
                : ' a-quiz-answer--incorrect';
        }

        return (
            <label htmlFor={`answer-${id}`} className={labelClassNames}>
                {questionType === 'single-choice' ? (
                    <RadioButton />
                ) : (
                    <Checkbox />
                )}
                {title}
            </label>
        );
    }
);

export default Answer;
