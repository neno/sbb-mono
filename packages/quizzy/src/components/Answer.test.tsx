import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Answer from './Answer';

const questionId = '100';
let questionType = 'single-choice';
const id = '101';
const title = 'Option 1';
const toggleAnswer = jest.fn(() => {});
let correct = false;
let showResults = false;
let checked = false;

const getAnswerComponent = () => (
    <Answer
        id={id}
        title={title}
        correct={correct}
        questionId={questionId}
        questionType={questionType}
        showResults={showResults}
        toggleAnswer={toggleAnswer}
        checked={checked}
    />
);

describe('Answer', () => {
    afterEach(() => {
        showResults = false;
        correct = false;
        checked = false;
    });

    test('toggleAnswer', () => {
        render(getAnswerComponent());
        fireEvent.click(screen.getByLabelText(title));
        expect(toggleAnswer).toHaveBeenCalled();
    });

    test('Not checked', () => {
        render(getAnswerComponent());
        const input = screen.getByLabelText(title);
        expect(input.getAttribute('checked')).toBeNull();
    });

    test('Checked', () => {
        checked = true;
        render(getAnswerComponent());
        const input = screen.getByLabelText(title);
        expect(input.getAttribute('checked')).not.toBeNull();
    });

    describe('Atomic Design and BEM', () => {
        test('Default', () => {
            render(getAnswerComponent());
            expect(screen.getByText(title).classList).toContain(
                'a-quiz-answer'
            );
            expect(screen.getByText(title).classList).not.toContain(
                'a-quiz-answer--incorrect'
            );
            expect(screen.getByText(title).classList).not.toContain(
                'a-quiz-answer--correct'
            );
        });

        describe('Show Results', () => {
            test('For incorrect answer', () => {
                showResults = true;
                correct = false;
                render(getAnswerComponent());
                expect(screen.getByText(title).classList).toContain(
                    'a-quiz-answer--incorrect'
                );
            });

            test('For correct answer', () => {
                showResults = true;
                correct = true;
                render(getAnswerComponent());
                expect(screen.getByText(title).classList).toContain(
                    'a-quiz-answer--correct'
                );
            });
        });
    });

    describe('Single-Choice', () => {
        beforeEach(() => {
            questionType = 'single-choice';
        });

        test('Snapshot', () => {
            const { container } = render(getAnswerComponent());
            expect(container.firstChild).toMatchSnapshot();
        });

        test('Radio button', () => {
            render(getAnswerComponent());
            const input = screen.getByLabelText(title);
            expect(questionType).toBe('single-choice');
            expect(input.getAttribute('type')).toBe('radio');
            expect(input.getAttribute('name')).toBe(`answer-${questionId}`);
        });
    });

    describe('Multiple-Choice', () => {
        beforeEach(() => {
            questionType = 'multiple-choice';
        });

        test('Snapshot', () => {
            const { container } = render(getAnswerComponent());
            expect(container.firstChild).toMatchSnapshot();
        });

        test('Checkbox', () => {
            render(getAnswerComponent());
            const input = screen.getByLabelText(title);
            expect(input.getAttribute('type')).toBe('checkbox');
            expect(input.getAttribute('name')).toBe(`answer-${id}`);
        });
    });
});
