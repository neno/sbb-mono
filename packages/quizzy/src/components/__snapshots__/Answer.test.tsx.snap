// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer Multiple-Choice Snapshot 1`] = `
<label
  class="a-quiz-answer a-quiz-answer--multiple-choice"
  for="answer-101"
>
  <input
    id="answer-101"
    name="answer-101"
    type="checkbox"
  />
  Option 1
</label>
`;

exports[`Answer Single-Choice Snapshot 1`] = `
<label
  class="a-quiz-answer a-quiz-answer--single-choice"
  for="answer-101"
>
  <input
    id="answer-101"
    name="answer-100"
    type="radio"
  />
  Option 1
</label>
`;
